flow360/plugins/report/report_items.py

```diff
@@ -890,23 +988,95 @@ def _calculate_focus_y_limits(
             end_idx = int(len(xs_np) * end_frac)
 
             if end_idx > start_idx:
-                focused_y = ys_np[start_idx:end_idx]
+                subset_y = ys_np[start_idx:end_idx]
             else:
-                focused_y = ys_np
+                subset_y = ys_np
+
+            if len(subset_y) > 0:
+                all_subset_y.extend(subset_y.tolist())
+
+        return all_subset_y
+
+    def _calculate_y_min_max(
+        self, all_subset_y, type_name: Literal["SubsetLimit", "FixedRangeLimit"]
+    ) -> Tuple[float, float]:
+        """
+        Given a subset of data and ylim type,
+        calculate min and max y values.
+        """
+        subset_y_min = float(min(all_subset_y))
+        subset_y_max = float(max(all_subset_y))
 
-            if len(focused_y) > 0:
-                all_focused_y.extend(focused_y.tolist())
+        if type_name == "SubsetLimit":
+            y_range = subset_y_max - subset_y_min
+            y_min = subset_y_min - self.ylim.offset * y_range
+            y_max = subset_y_max + self.ylim.offset * y_range
 
-        if not all_focused_y:
-            return None, None
+        else:
+            y_center = (subset_y_max + subset_y_min) / 2
+            y_min = y_center - 0.5 * self.ylim.fixed_range
+            y_max = y_center + 0.5 * self.ylim.fixed_range
 
-        global_y_min = float(min(all_focused_y))
-        global_y_max = float(max(all_focused_y))
-        y_range = global_y_max - global_y_min
-        y_min = global_y_min - 0.25 * y_range
-        y_max = global_y_max + 0.25 * y_range
         return (y_min, y_max)
 
+    def _calculate_ylimits(
+        self, x_series_list: List[List[float]], y_series_list: List[List[float]]
+    ) -> Tuple[Optional[float], Optional[float]]:
+        """
+        Calculate ylim based on provided input.
+        """
+        ylim = self.ylim
+
+        if ylim is None:
+            return None
+
+        if isinstance(ylim, Tuple):
+            return ylim
+
+        if isinstance(ylim, ManualLimit):
+            y_range = (ylim.lower, ylim.upper)
+
+        elif isinstance(ylim, SubsetLimit):
+            start_frac, end_frac = ylim.subset
+            type_name = ylim.type_name
+
+            all_subset_y = self._calculate_subset(
+                x_series_list, y_series_list, start_frac, end_frac
+            )
+
+            if not all_subset_y:
+                return (None, None)
+
+            y_range = self._calculate_y_min_max(all_subset_y, type_name)
+
+        else:
+            type_name = ylim.type_name
+
+            if ylim.center_strategy == "last":
+                all_last_y = []
+                for ys in y_series_list:
+                    last_y = ys[-1]
+                    all_last_y.append(last_y)
+
+                if not all_last_y:
+                    return (None, None)
+
+                y_range = self._calculate_y_min_max(all_last_y, type_name)
```

Comment:
these lines are repeated in else statement. The only difference is name of variable, I think it can be taken out of if statement